src/track/globaltrackcache.cpp

```diff
@@ -30,240 +36,356 @@ TrackRef createTrackRef(const Track& track) {
                 return false;
             }
-            VERIFY_OR_DEBUG_ASSERT((*j).ref == trackRef) {
+            VERIFY_OR_DEBUG_ASSERT((*j).second == plainPtr) {
                 return false;
             }
         }
     }
-    for (TracksByCanonicalLocation::const_iterator i(m_tracksByCanonicalLocation.begin()); i != m_tracksByCanonicalLocation.end(); ++i) {
-        const TrackRef trackRef((*i).ref);
-        const TrackId trackId(trackRef.getId());
-        const QString canonicalLocation(trackRef.getCanonicalLocation());
-        VERIFY_OR_DEBUG_ASSERT(!canonicalLocation.isEmpty()) {
+    for (TracksByCanonicalLocation::const_iterator i = m_tracksByCanonicalLocation.begin(); i != m_tracksByCanonicalLocation.end(); ++i) {
+        const Track* plainPtr = (*i).second;
+        VERIFY_OR_DEBUG_ASSERT(plainPtr) {
             return false;
         }
-        VERIFY_OR_DEBUG_ASSERT(i.key() == canonicalLocation) {
+        const TrackRef trackRef = createTrackRef(*plainPtr);
+        VERIFY_OR_DEBUG_ASSERT(!trackRef.getCanonicalLocation().isEmpty()) {
             return false;
         }
-        VERIFY_OR_DEBUG_ASSERT(createTrackRef(*(*i).plainPtr) == trackRef) {
+        VERIFY_OR_DEBUG_ASSERT(trackRef.getCanonicalLocation() == (*i).first) {
             return false;
         }
-        VERIFY_OR_DEBUG_ASSERT(1 == m_tracksByCanonicalLocation.count(canonicalLocation)) {
+        VERIFY_OR_DEBUG_ASSERT(1 == m_tracksByCanonicalLocation.count(trackRef.getCanonicalLocation())) {
             return false;
         }
-        TracksById::const_iterator j(
-                m_tracksById.find(trackId));
+        TracksById::const_iterator j = m_tracksById.find(trackRef.getId());
         VERIFY_OR_DEBUG_ASSERT(
-                (m_tracksById.end() == j) || ((*j).ref == trackRef)) {
+                (m_tracksById.end() == j) || ((*j).second == plainPtr)) {
             return false;
         }
     }
     return true;
 }
 
-GlobalTrackCacheLocker GlobalTrackCache::lookupById(
-        const TrackId& trackId) const {
-    GlobalTrackCacheLocker cacheLocker;
-    if (trackId.isValid()) {
-        const TrackPointer pTrack(lookupInternal(trackId));
-        if (pTrack) {
-            cacheLocker.m_lookupResult = GlobalTrackCacheLookupResult::HIT;
-            cacheLocker.m_trackRef = createTrackRef(*pTrack);
-            cacheLocker.m_pTrack = pTrack;
-        } else {
-            cacheLocker.m_lookupResult = GlobalTrackCacheLookupResult::MISS;
+void GlobalTrackCache::deactivate() {
+    DEBUG_ASSERT(verifyConsistency());
+    // Ideally the cache should be empty when destroyed.
+    // But since this is difficult to achieve all remaining
+    // cached tracks will evicted no matter if they are still
+    // referenced or not. This ensures that the eviction
+    // callback is triggered for all modified tracks before
+    // exiting the application.
+    while (!m_indexedTracks.empty()) {
+        evictAndDelete(
```

Comment:
We must not delete here, because there are references out there.

Reply:
Ups, yes, this is invalid now that the pointer is deleted unconditionally by the callback interfaces.